designs/disruption-controls-by-reason.md

```diff
@@ -0,0 +1,271 @@
+      // This is required if Schedule is set.
+      // This regex has an optional 0s at the end since the duration.String() always adds
+      // a 0s at the end.
+      // +kubebuilder:validation:Pattern=`^([0-9]+(m|h)+(0s)?)$`
+      // +kubebuilder:validation:Type="string"
+      // +optional
+      Duration *metav1.Duration `json:"duration,omitempty" hash:"ignore"`
+}
+
+```
+
+
+##### Example
+```yaml
+apiVersion: karpenter.sh/v1beta1
+kind: NodePool
+metadata:
+  name: default
+spec: # This is not a complete NodePool Spec.
+  disruption:
+    budgets:
+    - schedule: "* * * * *"
+      reason: [Drift]
+      nodes: 10
+    # For all other reasons , only allow 5 nodes to be disrupted at a time
+    - nodes: 5
+      schedule: "* * * * *"
+
+```
+
+In the original proposed spec, karpenter allows the user to specify up to [50 budgets](https://github.com/kubernetes-sigs/karpenter/blob/main/pkg/apis/v1beta1/nodepool.go#L96)
+
+If there are multiple active budgets, karpenter takes the most restrictive budget. This same principle will be applied to the disruption budgets in this approach. The only difference in behavior is that each window will apply to list of reasons that are specifed rather than just all disruption methods.
+
+### Pros + Cons 
+* 👍 No nested definitions required 
+* 👍👍 Extending existing budgets api. No Breaking API Changes, completely backwards compatible
+* 👍 Easy Top level list. Defining the list of reasons in one dimension makes it easier to share that configuration with other nodepools, and easier to apply one generic schedule to many disruption reasons
+* 👎 With reason being clearly tied to budgets, and other api logic being driven by disruption reason, we lose the chance to generalize per Reason controls 
+
+### Approach B: Defining Per Reason Controls  
+Ideally, we could move all generic controls that easily map into other reasons into one set of reason controls, this applies to budgets and other various disruption controls that could be more generic. 
+### Proposed Spec 
+```go
+type Disruption struct {
+    Default	  DisruptionSpec `json:defaults"`
+    Consolidation ConsolidationSpec `json:"consolidation"`
+    Drift         DriftSpec         `json:"drift"`
+    Expiration    ExpirationSpec    `json:"expiration"`
+    Emptiness     EmptinessSpec     `json:"emptiness"`
+}
+
+type DisruptionCommonSpec struct {
+    DisruptAfter string   `json:"disruptAfter"`
+    Budgets      []Budget `json:"budgets"`
+}
+
+type ConsolidationSpec struct {
+    DisruptionCommonSpec
+    ConsolidationPolicy string `json:"consolidationPolicy"`
+}
+
+type DriftSpec struct {
+    DisruptionCommonSpec
+}
+
+type ExpirationSpec struct {
+    DisruptionCommonSpec
+}
+
+type EmptinessSpec struct {
+    DisruptionCommonSpec
+}
+
+type Budget struct {
+    Nodes    string  `json:"nodes"`
+    Schedule *string `json:"schedule,omitempty"`
+    Duration *string `json:"duration,omitempty"`
+    Reasons []string 
```

Comment:
We have three options for the name of the field. 

1. Actions 
2. Reasons 
3. Methods 

I removed all the talk from the doc of drift reasons, and we aren't designing it right now, but also this API will be used directly after to represent them or other reasons. 


REasons to me makes more sense than Methods because its more extensible, and methods directly reference the primary 4 methods in the karpenter disruption controller.

Reply:
Rather than changing the API in the future, it would be good to see a general list of arbitrary reasons that can be extended. 

Allthough we may see that we don't want to reuse the same stucture for that design. Lets just call it methods and if we need to change it we can.

Reply:
Fwiw I'm +1 for "reasons". "Methods" doesn't read as well in plain English. It's essentially noun (reason=expired) vs verb (method=expire), and declarative API policies tend towards nouns.

Reply:
Lets have a vote at the next WG, I am leaning more on reasons as welll after reflecting a bit on it.

Reply:
Ah I thought reasons referred to a specific per-method reason of disrupting, like NodePoolDrift or Underutilized for Consolidation. We use methods in code, but I agree reasons semantically make more sense to refer to these. In docs we call these methods. Wherever we land on this, it'd be good to update that language throughout (obviously where it makes sense)

Reply:
Ah good call

Reply:
> Wherever we land on this, it'd be good to update that language throughout

FWIW, I think that users probably don't care about method when there's an overlap between consolidation and emptiness. I really only care, "is my node fully empty or only partially empty". There's perhaps an argument to be made that we should change our metrics and logging to align with "reason" in the new API rather than the other way around.